stack.yaml: switch from Stackage LTS to Nightly

peti opened the pull request:
Let's use a moderately recent compiler like ghc-8.10.3
instead of the old 8.8.x branch for development. Same goes
for our dependencies; development should take place with
recent versions of our dependencies.

This PR also fixes the build of `generatemanpage` and enables
it in our CI environment.

peti committed:
stack.yaml: switch from Stackage LTS to Nightly

Let's use a moderately recent compiler like ghc-8.10.3
instead of the old 8.8.x branch for development. Same goes
for our dependencies; development should take place with
recent versions of our dependencies.

liskin commented:
Doesn't seem like it actually fixes anything, does it?

And it doesn't help the Debian folks one bit either. :-(

peti commented:
> Doesn't seem like it actually fixes anything, does it?

Why? Have you tried running `stack build` with that PR applied? Did it not work?

liskin commented:
> Why? Have you tried running `stack build` with that PR applied? Did it not work?

I haven't, personally, but GitHub Actions have.

peti commented:
You should really try it yourself. You might learn something.

peti committed:
Run CI builds with Stackage Nightly.

Unfortunately, our CI builds have a hard-coded assumption that all kinds of
LTS-X builds can be performed with the same stack.yaml file, which is not true.
We need to use specialized build instructions for each version of the
environment we're building in, e.g. stack-lts-16.yaml, stack-nightly.yaml, and
so on. Making that change is not super hard, but it's not completely trivial
either because the hooks use a hash of the stack file to determine whether the
cash is still valid, etc., and I'm not entirely sure about the best way to
adapt that code to support multiple YAML files.

For the time being, the most important build IMHO is Stackage Nightly,
because (a) we really want to support the nightly snapshot since (b) it can
compile the genmanpage executable relatively easily.

peti commented:
Now CI builds fine, too.

slotThe commented:
@peti Why was this merged?  As far as I can tell, no consensus was reached regarding which approach (this or #260) is to be favoured.  

Only building with stack nightly is highly problematic in its own right—we, for example, have [this](https://github.com/xmonad/xmonad/blob/master/xmonad.cabal#L30) in our cabal file, promising to support GHC versions `8.4.4` and up; not just `8.10.x`.  Also [this](https://github.com/xmonad/xmonad/blob/master/xmonad.cabal#L90) way of detecting unused imports is now not checked in the CI; does it now fall onto the developer to check this manually?    

I'm voting to revert this until a proper consensus has been reached.

liskin commented:
@peti Seriously? This fixes none of the problems and breaks most of the testing.

@byorgey @geekosaur Can you guys please step in?

peti commented:
> @peti Why was this merged?

This was merged, because as a release manager for xmonad I want to be sure that the tool to generate our man page works, and the previous solution did not build it. I also want to make sure that we support *recent* compilers, whereas the previous solution focussed on old ones.

Having said that, I am perfectly happy to re-enable those old LTS builds asap. Please refer to https://github.com/xmonad/xmonad/pull/261/commits/3015968ee43a71648c260857d73dba19168083fe for details.
